testHelpers/events.js

```diff
@@ -0,0 +1,89 @@
+const ethers = require('hardhat').ethers;
+
+const eventNames = {
+  LOG_ORDER_CREATED: 'LogOrderCreated',
+  LOG_PROMISE_CREATED: 'LogPromiseCreated',
+  LOG_VOUCHER_DELIVERED: 'LogVoucherDelivered',
+  LOG_VOUCHER_REDEEMED: 'LogVoucherRedeemed',
+  LOG_VOUCHER_REFUNDED: 'LogVoucherRefunded',
+  LOG_VOUCHER_COMPLAIN: 'LogVoucherComplain',
+  LOG_VOUCHER_FAULT_CANCEL: 'LogVoucherFaultCancel',
+  LOG_EXPIRATION_TRIGGERED: 'LogExpirationTriggered',
+  LOG_FINALIZED_VOUCHER: 'LogFinalizeVoucher',
+  LOG_AMOUNT_DISTRIBUTION: 'LogAmountDistribution',
+  LOG_FUNDS_RELEASED: 'LogFundsReleased',
+  LOG_WITHDRAWAL: 'LogWithdrawal',
+  LOG_COMPLAIN_PERIOD_CHANGED: 'LogComplainPeriodChanged',
+  LOG_CANCEL_FAULT_PERIOD_CHANGED: 'LogCancelFaultPeriodChanged',
+  TRANSFER_SINGLE: 'TransferSingle',
+  TRANSFER: 'Transfer',
+  TRANSFER_BATCH: 'TransferBatch',
+  LOG_DISASTER_STATE_SET: 'LogDisasterStateSet',
+  LOG_WITHDRAW_TOKENS_ON_DISASTER: 'LogWithdrawTokensOnDisaster',
+  LOG_WITHDRAW_ETH_ON_DISASTER: 'LogWithdrawEthOnDisaster',
+  LOG_ETH_LIMIT_CHANGED: 'LogETHLimitChanged',
+  LOG_TOKEN_LIMIT_CHANGED: 'LogTokenLimitChanged',
+  APPROVAL_FOR_ALL: 'ApprovalForAll',
+  APPROVAL: 'Approval',
+  LOG_BR_SET: 'LogBosonRouterSet',
+  LOG_ERC1155_ERC721_SET: 'LogTokenContractSet',
+  LOG_VK_SET: 'LogVoucherKernelSet',
+  LOG_CASHIER_SET: 'LogCashierSet',
+};
+
+function getEventArgsFromFactory(factory, eventName) {
+  let [eventFragment] = factory.interface.fragments.filter(
+    (e) => e.name == eventName
+  );
+  return eventFragment.inputs.map((e) => e.name);
+}
+
+function getEventArgTypesFromFactory(factory, eventName) {
+  let [eventFragment] = factory.interface.fragments.filter(
+    (e) => e.name == eventName
+  );
+  return eventFragment.inputs
+    .filter((e) => e.indexed != true)
+    .map((e) => e.type);
+}
+
+function assertEventEmitted(receipt, factory, eventName, callback) {
+  let found = false;
+
+  let eventFragment = factory.interface.fragments.filter(
+    (e) => e.name == eventName
+  );
+  const interface = new ethers.utils.Interface(eventFragment);
+
+  for (const log in receipt.logs) {
+    const topics = receipt.logs[log].topics;
+    for (const index in topics) {
+      const encodedTopic = topics[index];
+
+      try {
+        // CHECK IF TOPIC CORRESPONDS TO THE EVENT GIVEN TO FN
+        let event = interface.getEvent(encodedTopic);
+
+        if (event.name == eventName) {
+          found = true;
+          const eventArgs = interface.parseLog(receipt.logs[log]).args;
+          callback(eventArgs);
+        }
+      } catch (e) {
+        if (e.message.includes('no matching event')) continue;
+        throw new Error(e);
+      }
+    }
+  }
+
+  if (!found) {
+    throw new Error(`Event with name ${eventName} was not emitted!`);
+  }
```

Comment:
Isn't using something like 

```
const blockNumber = await ethers.provider.getBlockNumber();
const events = await instance.queryFilter("LogTokenOwnerAddressChanged", blockNumber - 1, blockNumber);
```

a more straight-forward way to get the event? The `instance` in this case is the instance the original function was called on. When using this manner of getting events, I've found I have to check the current block and the one before. Unlike with truffle tests (which usually process transactions from a test case in one block), with hardhat it seems there can be multiple block.

Reply:
By default Hardhat Network is configured to auto-mine blocks, immediately upon receiving each transaction unless there is no specific configuration to change that behaviour - [reference](https://hardhat.org/hardhat-network/explanation/mining-modes.html). I find this way more robust as it does not rely on a block, but on the transaction receipt itself. This way we would not need to know which block this has been included at.  Other way around we might have to guess / increase range on the query filter in case of change of configs